topology/probes/istio/virtualservice.go

```diff
@@ -49,3 +52,129 @@ func (h *virtualServiceHandler) Dump(obj interface{}) string {
 func newVirtualServiceProbe(client interface{}, g *graph.Graph) k8s.Subprobe {
 	return k8s.NewResourceCache(client.(*kiali.IstioClient).GetIstioNetworkingApi(), &kiali.VirtualService{}, "virtualservices", g, &virtualServiceHandler{})
 }
+
+type virtualServicePodLinker struct {
+	graph      *graph.Graph
+	vsCache    *k8s.ResourceCache
+	podCache   *k8s.ResourceCache
+	vsIndexer  *graph.Indexer
+	podIndexer *graph.Indexer
+}
+
+type virtualServiceInfo struct {
+	App     string `mapstructure:"host"`
+	Version string `mapstructure:"subset"`
+}
+
+func (i *virtualServiceInfo) getIndexerValues() []interface{} {
+	values := []interface{}{i.App}
+	if i.Version != "" {
+		values = append(values, i.Version)
+	}
+	return values
+}
+
+type virtualServiceSpec struct {
+	HTTP []struct {
+		Route []struct {
+			Destination virtualServiceInfo `mapstructure:"destination"`
+		} `mapstructure:"route"`
+	} `mapstructure:"http"`
+}
+
+func getVirtualServiceInfos(vs *kiali.VirtualService) ([]virtualServiceInfo, error) {
+	vsSpec := &virtualServiceSpec{}
+	if err := mapstructure.Decode(vs.Spec, vsSpec); err != nil {
+		return nil, err
+	}
+	var infos []virtualServiceInfo
+	for _, http := range vsSpec.HTTP {
+		for _, route := range http.Route {
+			infos = append(infos, route.Destination)
+		}
+	}
+	return infos, nil
+}
+
+func (vspl *virtualServicePodLinker) GetABLinks(vsNode *graph.Node) (edges []*graph.Edge) {
+	vs := vspl.vsCache.GetByNode(vsNode)
+	if vs, ok := vs.(*kiali.VirtualService); ok {
+		infos, _ := getVirtualServiceInfos(vs)
+		for _, info := range infos {
+			podNodes, _ := vspl.podIndexer.Get(info.getIndexerValues()...)
+			for _, podNode := range podNodes {
+				id := graph.GenID(string(vsNode.ID), string(podNode.ID), "RelationType", "virtualservice")
+				edges = append(edges, vspl.graph.NewEdge(id, vsNode, podNode, nil, ""))
+			}
+		}
+	}
+	return
+}
+
+func (vspl *virtualServicePodLinker) GetBALinks(podNode *graph.Node) (edges []*graph.Edge) {
+	pod := vspl.podCache.GetByNode(podNode)
+	if pod, ok := pod.(*v1.Pod); ok {
+		infoWithVersion := virtualServiceInfo{App: pod.Labels["app"], Version: pod.Labels["version"]}
+		infoWithoutVersion := virtualServiceInfo{App: pod.Labels["app"], Version: ""}
+		vsWithVersionNodes, _ := vspl.vsIndexer.Get(infoWithVersion.getIndexerValues()...)
+		vsWithoutVersionNodes, _ := vspl.vsIndexer.Get(infoWithoutVersion.getIndexerValues()...)
+		vsNodes := append(vsWithVersionNodes, vsWithoutVersionNodes...)
+		for _, vsNode := range vsNodes {
+			id := graph.GenID(string(vsNode.ID), string(podNode.ID), "RelationType", "virtualservice")
+			edges = append(edges, vspl.graph.NewEdge(id, vsNode, podNode, nil, ""))
+		}
+	}
+	return
+}
+
+func newVirtualServicePodLinker(g *graph.Graph, probes map[string]k8s.Subprobe) probe.Probe {
+	vsProbe := probes["virtualservice"]
+	podProbe := k8s.Subprobes["pod"]
```

Comment:
This assumes the `k8s` probe is enabled, something we tried to avoid.
We should do:
`podCache := k8s.NewResourceCache(....)`

Reply:
we don't want "pod" objects to appear in  the graph in the case that the k8s probe is disabled ... in that case we would expect that "virtualservice" objects would float about (not connected to their respective "namespace" objects), and we would not expect to see any links between "virtualservice" objects and "pod" object (as the later should not appear in the graph). ... so attempting to create a functional pod Subprobe in this case would be incorrect. (basically the ResourceCache listens to k8s and creates graph elements ...).

So back to exporting and using k8s.Subprobes ... this means that *only* if the resource/subprobe already exists - then it can be used - if it does not exist we get a `nil` return and exit the `newVirtualServicePodLinker()` as expected.

As to the proposed implementation of k8s.Subprobes ... I would propose making this uniform for both k8s and istio possibly using the following API (implemented in k8s/probe.go):
- k8s.PutSubprobe(manager, name string, subprobe k8s.Subprobe)
- k8s.GetSubprobe(manager, name string) k8s.Subprobe
- k8s.ListSubprobes(manager string) []k8s.Subprobe

the internal implementation could be a map[string]map[string]Subprobe

Reply:
Ok. I guess accessing Subprobes works for now